### PR TITLE
[WIP] Added range options to --scale

### LINKED ISSUE #822: support --scale
I find myself often needing to do: `--min-scale=X --max-scale=x`, usually when x=1.

It might be nice to support a short-cut here: `--scale=x` to mean "set both min & max".

Just a thought...

### PR DESCRIPTION
**Pull request by mpetason**
## Description

Adds range options to --scale. 

## Changes

* Adds three more options to --scale (1..5, 1.., ..5)

## Reference

Fixes #822

### CODE REVIEW

**Comment by maximilien** [comment]
/ok-to-test

Ready for some reviews @mpetason?

**Comment by mpetason** [comment]
Not ready for a review yet. I need to cleanup the range section and create a helper function. Right now it's just in an "it works" status and I wanted to show some work.

File: docs/cmd/kn_service_create.md [hunk 1]
```diff
@@ -86,7 +86,7 @@ kn service create NAME --image IMAGE
       --requests-cpu string           DEPRECATED: please use --request instead. The requested CPU (e.g., 250m).
       --requests-memory string        DEPRECATED: please use --request instead. The requested memory (e.g., 64Mi).
       --revision-name string          The revision name to set. Must start with the service name and a dash as a prefix. Empty revision name will result in the server generating a name for the revision. Accepts golang templates, allowing {{.Service}} for the service name, {{.Generation}} for the generation, and {{.Random [n]}} for n random consonants. (default "{{.Service}}-{{.Random 5}}-{{.Generation}}")
-      --scale int                     Minimum and maximum number of replicas.
+      --scale string                  Minimum and maximum number of replicas. Example: --scale 5 or --scale 1..5 or --scale 1.. or --scale ..5.
```
Comment (maximilien) [inline docs/cmd/kn_service_create.md:89]: What does it mean to do —scale ..5 (up to five?). Maybe lets be clear with text...

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 2]
```diff
@@ -82,7 +83,7 @@ func (p *ConfigurationEditFlags) addSharedFlags(command *cobra.Command) {
 	command.Flags().MarkHidden("max-scale")
 	p.markFlagMakesRevision("max-scale")
 
-	command.Flags().IntVar(&p.Scale, "scale", 0, "Minimum and maximum number of replicas.")
+	command.Flags().StringVar(&p.Scale, "scale", "", "Minimum and maximum number of replicas.")
```
Comment (maximilien) [inline pkg/kn/commands/service/configuration_edit_flags.go:86]: Is “” a valid value? Perhaps use default?

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 3]
```diff
@@ -333,13 +334,45 @@ func (p *ConfigurationEditFlags) Apply(
...
 		} else {
-			err = servinglib.UpdateMaxScale(template, p.Scale)
-			if err != nil {
-				return err
-			}
-			err = servinglib.UpdateMinScale(template, p.Scale)
-			if err != nil {
-				return err
+			if !strings.Contains(p.Scale, "..") {
+				scaleInt, _ := strconv.Atoi(p.Scale)
```
Comment (maximilien) [inline pkg/kn/commands/service/configuration_edit_flags.go:338]: Why are you ignoring the `Atoi(...)` error? Just catch and check it

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 4]
```diff
@@ -333,13 +334,45 @@ func (p *ConfigurationEditFlags) Apply(
...
+				if err != nil {
+					return err
+				}
+				err = servinglib.UpdateMinScale(template, scaleInt)
+				if err != nil {
+					return err
+				}
+			} else if len(p.Scale) == 4 {
+				scaleParts := strings.Split(p.Scale, "..")
+				scaleMin, _ := strconv.Atoi(scaleParts[0])
```
Comment (maximilien) [inline pkg/kn/commands/service/configuration_edit_flags.go:349]: Same here... don’t ignore errors

**Comment by rhuss** [comment]
fyi, the fix for the failing CI is included as part of #902

**Comment by maximilien** [comment]
Wonder if this is a flake

/test pull-knative-client-build-tests

**Comment by rhuss** [comment]
@mpetason you would need to recreate the docs via `build.sh` and commit the generated/updated markdown files

**Comment by mpetason** [comment]
@rhuss will do. I should have an update for this today.

**Comment by mpetason** [comment]
Still WIP, need to update tests and fix how I'm checking for errors.

**Comment by mpetason** [comment]
@maximilien @rhuss I think I'm ready for a review now. I need to update the docs (reverted back with the last build), but would like some pointers on changes I can make to the code/tests.

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 5]
```diff
@@ -82,7 +84,7 @@ func (p *ConfigurationEditFlags) addSharedFlags(command *cobra.Command) {
 	command.Flags().MarkHidden("max-scale")
 	p.markFlagMakesRevision("max-scale")
 
-	command.Flags().IntVar(&p.Scale, "scale", 0, "Minimum and maximum number of replicas.")
+	command.Flags().StringVar(&p.Scale, "scale", "1", "Minimum and maximum number of replicas.")
```
Comment (daisy-ycguo) [inline pkg/kn/commands/service/configuration_edit_flags.go:87]: I suggest to add a sample at the end of the explanation, e.g.
```
Minimum and maximum number of replicas. (e.g., 1..5).
```
in order to help end users understand the format clearly.

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 6]
```diff
@@ -480,3 +486,32 @@ func (p *ConfigurationEditFlags) AnyMutation(cmd *cobra.Command) bool {
 	}
 	return false
 }
+
+func (p *ConfigurationEditFlags) ScaleConversion(scale string) (scaleMin int, scaleMax int, err error) {
```
Comment (daisy-ycguo) [inline pkg/kn/commands/service/configuration_edit_flags.go:490]: Do you want the method `ScaleConversion` to be visible outside package? If the method is not used by any other packages, I suggest to change the method name to start with a lower case letter, e.g. `scaleConversion`

**Review by daisy-ycguo — CHANGES_REQUESTED** [review]
Thank you for the contributions. I think this PR looks good generally. I just left a few comments.

**Comment by mpetason** [comment]
I didn't rebase this correctly and have no clue how to fix it. I guess I need to close/delete this PR and try again?

**Comment by mpetason** [comment]
Can't figure out how to fix the issue where there are so many files added. Tried a rebase but it doesn't look like that helped. Going to close this out for good and open a new PR with the same work.

**Comment by mpetason** [comment]
Figured out how to clean this up after rebase. I'm probably going to close this out and re-open since I'm still running into conflicts with the command information files.

**Pull request closed without merging**